Upgrade libp2p

ryanio opened the pull request:
This PR upgrades all `libp2p` packages to latest. These updates include:

* deprecated `peer-info` module refactored out
  * `PeerId` and `multiaddr`used  instead
* `libp2p` callbacks converted to promises
* `bootnode` type refactored out
  * `multiaddr` used instead
* libp2p connection encryption: `secio` (deprecated) swapped for `noise`
* `pull-stream` refactored to [async iterables](https://github.com/libp2p/js-libp2p/blob/master/doc/STREAMING_ITERABLES.md)
* various type improvements
* misc. fixes:
  * `package.json`: fix `ethereumjs` bin path
  * `cli.ts`: use `args.datadir` for `datadir`
  *  `browser/index.ts`: fix init logic

ryanio commented:
ci is having some trouble locating the types for `cids` during build:

`Error: @ethereumjs/client: ../../node_modules/libp2p/dist/src/content-routing.d.ts(86,19): error TS2307: Cannot find module 'cids/src/' or its corresponding type declarations.`

I'm having trouble reproducing locally, everything is passing for me on both node v10 and v12.

The path shouldn't have `/src/` in it, but it seems auto-generated somehow. The `@typedef` that it's generated from seems correct: [`* @typedef {import('cids')} CID`](https://github.com/libp2p/js-libp2p/blob/58d4f9a91542f180907689319d159b326f71096d/src/content-routing.js#L12). It could be an artifact from their build tool `aegir` that generates the `.d.ts` files.

Looking into this I did find that the latest `cids` release ([v1.1.4](https://github.com/multiformats/js-cid/commit/f5839b9788abe231e8d563d64f0fe3db61363c5c)) fixed the package.json `types` location, and in the past ([v1.0.2](https://github.com/multiformats/js-cid/blob/v1.0.2/src/index.d.ts)) there was an `index.d.ts` file in the `src` folder.

evertonfraga commented:
@ryanio I managed to reproduce it locally. Using either node 12 or 14, and the following commands:
```
npm run clean
npm i
```

ryanio commented on packages/client/package.json:
```diff
@@ -23,7 +23,7 @@
     "lint": "ethereumjs-config-lint",
     "lint:fix": "ethereumjs-config-lint-fix",
-    "tape": "tape -r ts-node/register",
+    "tape": "TS_NODE_FILES=true tape -r ts-node/register",
```
interestingly enough I found this was needed for `tape -r ts-node/register` to include the `typings` folder as defined in `tsconfig.json:include`

holgerd77 replied:
Ugh. 😄 

👍

holgerd77 commented:
What kind of practical tests did you run here? Is this working in a live environment syncing between two client instances following the browser example from the README e.g.? Or is this at foremost passing the test suite?

ryanio commented:
> What kind of practical tests did you run here? Is this working in a live environment syncing between two client instances following the browser example from the README e.g.? Or is this at foremost passing the test suite?

I am just now running the Readme Example 1 and ran into a few things.

First, the datadir arg didn't seem to be passed on correctly (should be fixed with https://github.com/ethereumjs/ethereumjs-vm/pull/1027/commits/fa1b08e8b7d0344bc1bab95c5b55c44aaab213db, although I am not sure if you wanted to use `~/Library/Ethereum/ethereumjs` or `./datadir`)

Second, I didn't get the PeerId hash at the end of the multiaddr listener address (`/ipfs/#`), instead it simply looks like:

`INFO [01-07|18:10:00] Listener up transport=libp2p url=/ip4/127.0.0.1/tcp/50507/ws`

This is a side effect of `peer-info` being refactored out. Should we emit the PeerId? For now I've added it back in with https://github.com/ethereumjs/ethereumjs-vm/pull/1027/commits/842e2047d618b4022c880cd4f06c0d98276d3f65.

Third, our `parseBootnodes` function isn't parsing a multiaddr format correctly. It looks like it only supports parsing `ip:port`, so I'm not sure how the example was working before! (although I know it was because I ran it before, I haven't gone back to master yet to see how it is currently). I added some commented out tests that fail in https://github.com/ethereumjs/ethereumjs-vm/pull/1027/commits/91eac53dcac9395bcd2d97600440709c135c92c3#diff-63ba7a6da3a637cc57cea0acbefcd815f8b13e4bb8698fce4e89f55a485a35af

Fourth, my light instance is not connecting to the full node for Example 1, I am also investigating this after fixing some of the above issues.

After I am able to get all this figured out I will also ensure that Example 2 for the browser also works :)

ryanio commented:
Still working on this, I will push up a new batch of commits soon.

I added a commit for `parseBootnodes` to additionally handle multiaddrs.

And I can confirm the peerId identifier in the multiaddr string `/ipfs/${peerId}` is now referenced as `/p2p/${peerId}`, so I will update references accordingly for that.

holgerd77 commented:
Ok. 😄 

For the data directory I think this should be networking-protocol (RLPx orLibp2p) agnostic so this should likely just take whatever is passed to `Config.dataDir` being it `~/Library/Ethereum/ethereumjs` or `./datadir`.

holgerd77 commented on packages/client/bin/cli.ts:
```diff
@@ -149,7 +148,7 @@ async function run() {
     syncmode: args.syncmode,
     lightserv: args.lightserv,
-    datadir: `${os.homedir()}/Library/Ethereum/ethereumjs`,
+    datadir: args.datadir,
```
Ah, can we just leave this on the old value (maybe my answer to your question was a bit misleading here)? Think this was a reasonable default and ensures that the data directory from "serious" client runs is different from the one used for the tests. Otherwise the test runs would override an already downloaded longer chain from the normal CLI runs.

Or do you have some other specific reason for this change?

If you are not satisfied that this is so os-specific another option would be to tackle https://github.com/ethereumjs/ethereumjs-vm/issues/999 opened by Paul and add more appropriate os-specific default directories. Maybe we should add this anyhow to the list of TODOs before a first more public release.

Yet another very simple option to avoid conflicts with the test datadir: we could add a separate default with a relative directory path (to already make it somewhat more os-friendly) directly here.

ryanio replied:
ah ok hm, well the reason I targeted this change was because the cli `--datadir` arg wasn't being used, so all client runs would use this default of `${os.homedir()}/Library/Ethereum/ethereumjs`, and I couldn't run two clients at once.

I see your point about a safe datadir for test suite runs though. I will fix.
My current thinking is to change:

```
    datadir: {
      describe: 'Data directory for the blockchain',
      default: Config.DATADIR_DEFAULT,
    },
```

to `default: ${os.homedir()}/Library/Ethereum/ethereumjs` and that should fix this problem, so a `--datadir` arg can override.

Then `Config.DATADIR_DEFAULT` can stay as `./datadir` for safe test executions.

How does that sound?

holgerd77 replied:
Yeah, sounds good! 😀

ryanio replied:
I have added this change, will come along next push that also rebases on latest master.

So `cli.ts` default datadir arg will be `${os.homedir()}/Library/Ethereum/ethereumjs`, which can be modified on start with `--datadir` flag.

And test runs should continue to use `./datadir` (`Config.DATADIR_DEFAULT` in `config.ts` when `options.datadir` is undefined).

holgerd77 commented:
What is the state of this?

ryanio commented:
I will work on getting this branch merge ready today.

ryanio commented:
This is now ready for final review.

To test libp2p local sync from a full client to a light client, you can follow the example in the readme, or these are the commands I was using:

First client:

```
npm run client:start -- --syncmode full --lightserv true  --datadir first --network rinkeby --transports rlpx libp2p:multiaddrs=/ip4/127.0.0.1/tcp/50505/ws
```

Look for the line:

```
INFO [02-11|16:37:11] Listener up transport=libp2p url=/ip4/127.0.0.1/tcp/50505/ws/p2p/QmRz3scEDzBBEWFFTHAeLpqhpjTNGigCVSyCtHvsjPz6r3
```

copy the Peer Identifier (`Qm...`) and run the second client after some thousand blocks are downloaded:

```
npm run client:start -- --syncmode light --network rinkeby --datadir second --transports libp2p:multiaddrs=/ip4/0.0.0.0/tcp/50506,bootnodes=/ip4/127.0.0.1/tcp/50505/ws/p2p/Qm...
```

holgerd77 reviewed (approved):
Thanks Ryan, this looks great! I couldn't give a too in-depth review since my knowledge here is too limited, but I skimmed closely over all the code parts. Not enough to spot local bugs but at least for getting a feeling on the overall changes and see if things make a consistent appearance and try to understand local refactoring and library update changes.

So I will approve here, think we do this similarly to the DNS PR from @cgewecke and continue to review this over time by using, since this is "only" touching the networking layer (in the worst case leading to a non-functioning connection, which is easily spotted) this should be ok.

I also tested locally with the example and then stumbled upon this `invalid transaction trie` thing mentioned in the chat, but this nevertheless at least shows that a connection is established, think we can do further higher-level protocol code changes subsequently.

So again, great work, cool, now we finally have got all of the client library in a consistent (for the very most part 😛 ) and up-to-date state and should have a great foundation for future work throughout the whole code base.

🎉
